sklearn/tree/_splitter.pyx

```diff
@@ -220,222 +221,178 @@ cdef class Splitter:
-        cdef double min_weight_leaf = self.min_weight_leaf
-        cdef UINT32_t* random_state = &self.rand_r_state
-
-        cdef SplitRecord best, current
-        cdef double current_proxy_improvement = -INFINITY
-        cdef double best_proxy_improvement = -INFINITY
-
-        cdef SIZE_t f_i = n_features
-        cdef SIZE_t f_j
-        cdef SIZE_t p
-        cdef SIZE_t i
-
-        cdef SIZE_t n_visited_features = 0
-        # Number of features discovered to be constant during the split search
-        cdef SIZE_t n_found_constants = 0
-        # Number of features known to be constant and drawn without replacement
-        cdef SIZE_t n_drawn_constants = 0
-        cdef SIZE_t n_known_constants = n_constant_features[0]
-        # n_total_constants = n_known_constants + n_found_constants
-        cdef SIZE_t n_total_constants = n_known_constants
-        cdef SIZE_t partition_end
-
-        _init_split(&best, end)
-
-        # Sample up to max_features without replacement using a
-        # Fisher-Yates-based algorithm (using the local variables `f_i` and
-        # `f_j` to compute a permutation of the `features` array).
-        #
-        # Skip the CPU intensive evaluation of the impurity criterion for
-        # features that were already detected as constant (hence not suitable
-        # for good splitting) by ancestor nodes and save the information on
-        # newly discovered constant features to spare computation on descendant
-        # nodes.
-        while (f_i > n_total_constants and  # Stop early if remaining features
-                                            # are constant
-                (n_visited_features < max_features or
-                 # At least one drawn features must be non constant
-                 n_visited_features <= n_found_constants + n_drawn_constants)):
-
-            n_visited_features += 1
-
-            # Loop invariant: elements of features in
-            # - [:n_drawn_constant[ holds drawn and known constant features;
-            # - [n_drawn_constant:n_known_constant[ holds known constant
-            #   features that haven't been drawn yet;
-            # - [n_known_constant:n_total_constant[ holds newly found constant
-            #   features;
-            # - [n_total_constant:f_i[ holds features that haven't been drawn
-            #   yet and aren't constant apriori.
-            # - [f_i:n_features[ holds features that have been drawn
-            #   and aren't constant.
-
-            # Draw a feature at random
-            f_j = rand_int(n_drawn_constants, f_i - n_found_constants,
-                           random_state)
-
-            if f_j < n_known_constants:
-                # f_j in the interval [n_drawn_constants, n_known_constants[
-                features[n_drawn_constants], features[f_j] = features[f_j], features[n_drawn_constants]
-
-                n_drawn_constants += 1
+ctypedef fused DataSplitterFused:
```

Comment:
Would a 3rd party package that inherits the splitting functions be able to theoretically use the `node_split_best` and `node_split_random` functions? 

E.g. if 1) `splitter` was an `ObliqueSplitter` subclass that inherits `Splitter` and 2) `data_splitter` was a `ObliqueDenseSplitter`? My understanding would suggest that `data_splitter` would make it unusable. Is that right?

Reply:
The data splitters themselves and `node_split_*` are not c-importable. A third party `ObliqueDenseSplitter` would not be able to use the `data_splitter` or the `node_split_{best|random}` functions.

This is the current status on `main`: only the Splitter is cimportable. Looking over your PR in https://github.com/scikit-learn/scikit-learn/pull/22754, I do not think making `node_split_*` cimportable would benefit it. A third party splitter would end up writing their own `node_split` function to manage the extra state.